dvc/remote/s3.py

```diff
@@ -250,90 +247,42 @@ def exists(self, path_infos):
 
         return results
 
-    def upload(self, from_infos, to_infos, names=None, no_progress_bar=False):
-        names = self._verify_path_args(to_infos, from_infos, names)
-
-        s3 = self.s3
-
-        for from_info, to_info, name in zip(from_infos, to_infos, names):
-            if to_info.scheme != "s3":
-                raise NotImplementedError
-
-            if from_info.scheme != "local":
-                raise NotImplementedError
-
-            logger.debug("Uploading '{}' to '{}'".format(from_info, to_info))
-
-            if not name:
-                name = from_info.name
-
-            total = os.path.getsize(fspath_py35(from_info))
-            cb = None if no_progress_bar else Callback(name, total)
-
-            try:
-                s3.upload_file(
-                    from_info.fspath,
-                    to_info.bucket,
-                    to_info.path,
-                    Callback=cb,
-                    ExtraArgs=self.extra_args,
-                )
-            except Exception:
-                msg = "failed to upload '{}'".format(from_info)
-                logger.exception(msg)
-                continue
+    @contextmanager
```

Comment:
why is this decorator needed?

Reply:
To use it in `with` statement. That is really needed only for ssh, to close connection, I use it for universality.